app/controllers/provider/admin/backend_apis_controller.rb

```diff
@@ -42,8 +42,8 @@ def update
   end
 
   def destroy
-    if @backend_api.services.empty? && @backend_api.destroy
-      redirect_to provider_admin_dashboard_path, notice: 'Backend API deleted'
+    if @backend_api.services.empty? && DeleteObjectHierarchyWorker.perform_later(@backend_api)
```

Comment:
The problem of doing it this way is from the moment it is requested to destroy to when it is actually destroyed, there can be a long time (maybe 5 minutes or more) and many things can happen in between:
1. It can be requested to destroy more than once, then that means that we will enqueue it more than once and if there are executed at the same time, we will make many of the things to destroy repeatedly at the same time... not very efficient...
2. It can be seen in the show view and in the index view, and it can be updated in the same time... maybe even with side effects. and it may even have traffic in the meantime...
3. If it fails and for some reason either it does not retry (for example we clean our redis for that deleting queue, which we have already done in the past in case of emergency) or it retries so many times that it won't retry anymore... we will have it kinda silently failed and the backend_api will keep being used normally... and in the backend_api we have nothing changed for us to even identify which backend_api aimed to be deleted.

So IMO we should do it as we are doing for a `service`, having a `state` that becomes `deleted` in the controller:
https://github.com/3scale/porta/blob/dd5c49fd0bc9903d3f6f88b1f0e0d98b4d2962ac/app/controllers/api/services_controller.rb#L64-L68
creating an event after the state has successfully changed:
https://github.com/3scale/porta/blob/dd5c49fd0bc9903d3f6f88b1f0e0d98b4d2962ac/app/models/service.rb#L200
https://github.com/3scale/porta/blob/dd5c49fd0bc9903d3f6f88b1f0e0d98b4d2962ac/app/models/service.rb#L614-L617

the event should contain only the info we need:
https://github.com/3scale/porta/blob/dd5c49fd0bc9903d3f6f88b1f0e0d98b4d2962ac/app/events/services/service_scheduled_for_deletion_event.rb#L1-L13

and then through a subscriber of the event, actually doing the deletion in background:
https://github.com/3scale/porta/blob/dd5c49fd0bc9903d3f6f88b1f0e0d98b4d2962ac/app/lib/event_store/repository.rb#L165

https://github.com/3scale/porta/blob/dd5c49fd0bc9903d3f6f88b1f0e0d98b4d2962ac/app/subscribers/service_deletion_subscriber.rb#L1-L12

Reply:
And I also find it simpler if we always do it the same way than if for every model we follow a different approach